common/cryptfs.c

```diff
@@ -307,12 +510,140 @@ create_device_node(const char *name)
+		ret = errno;
+		if (errno != ENXIO)
+			ERROR_ERRNO("Cannot remove dm-integrity device");
+		goto error;
+	}
+
+	/* remove device node if necessary */
+	char *device = cryptfs_get_device_path_new(name);
+	unlink(device);
+	mem_free(device);
+
+	DEBUG("Successfully deleted dm-integrity device");
+	/* We made it here with no errors.  Woot! */
+	ret = 0;
+
+error:
+	close(fd);
+	return ret;
+}
+
+/*
+static unsigned long
+get_provided_data_sectors(const char *real_blk_name)
+{
+	int fd;
+	unsigned long provided_data_sectors = 0;
+	char magic[8]; // "integrt" on a valid superblock
+
+	if ((fd = open(real_blk_name, O_RDONLY)) < 0) {
+		ERROR("Cannot open volume %s", real_blk_name);
+		return 0;
+	}
+
+	int bytes_read = read(fd, magic, sizeof(magic));
+	DEBUG("Bytes read: %d", bytes_read);
+	if (bytes_read != sizeof(magic)) {
+		ERROR("Cannot read superblock type from volume %s", real_blk_name);
+		goto errout;
+	}
+	if (strcmp(magic, "integrt") != 0) {
+		DEBUG("No existing integrity superblock detected on %s", real_blk_name);
+		provided_data_sectors = 1;
+		goto errout;
+	}
+
+	// 16 Bytes offset from start of superblock for provided_data_sectors
+	lseek(fd, 16, SEEK_SET);
+	bytes_read = read(fd, &provided_data_sectors, sizeof(provided_data_sectors));
+	DEBUG("Read bytes is: %d", bytes_read);
+
+	if (bytes_read != sizeof(provided_data_sectors) || provided_data_sectors == 0) {
+	        ERROR("Cannot read provided_data_sectors from volume %s", real_blk_name);
+		goto errout;
+	}
+
+errout:
+	DEBUG("Returning: provided_data_sectors= %ld",provided_data_sectors);
+	close(fd);
+	return provided_data_sectors;
+}
+*/
+
+//TODO: get number of data sectors available from superblock
+static unsigned long
+get_provided_data_sectors(const char *real_blk_name)
+{
+	int fd;
+	unsigned long provided_data_sectors = 0;
+	DEBUG("In get_provided_data_sectors");
+	if ((fd = open(real_blk_name, O_RDONLY)) < 0) {
+		ERROR("Cannot open volume %s", real_blk_name);
+		return 0;
+	}
+
+	DEBUG("Using 80%% of block device size as provided_data_sectors");
+	provided_data_sectors = 0.8 * get_blkdev_size(fd);
+
+	close(fd);
+	return provided_data_sectors;
+}
+
 char *
 cryptfs_setup_volume_new(const char *label, const char *real_blkdev, const char *key)
 //			 char *crypto_sys_path, unsigned int max_path)
 {
-	if (create_crypto_blk_dev(real_blkdev, key, label) < 0)
+	char *integrity_dev_label = mem_printf("%s-%s", label, "integrity");
+	unsigned long fs_size = get_provided_data_sectors(real_blkdev);
+	DEBUG("cryptfs_setup_volume_new");
+
+	if (fs_size <= 0) {
+		DEBUG("get_provided_data_sectors returned %lu!!", fs_size);
+		return NULL;
+	}
+
+	if (create_integrity_blk_dev(real_blkdev, integrity_dev_label, fs_size) < 0) {
+		DEBUG("create_integrity_blk_dev failed!");
 		return NULL;
+	}
+
+	char *integrity_dev = create_device_node(integrity_dev_label);
+	mem_free(integrity_dev_label);
+	if (!integrity_dev) {
+		ERROR_ERRNO("Could not create device node");
```

Comment:
ERROR, errno may not be set correctly (use ERRNO macros only after syscalls not for internal functions)

Reply:
See above